sklearn/utils/estimator_checks.py

```diff
@@ -360,55 +366,242 @@ def check_transformer_pickle(name, Transformer):
     assert_array_almost_equal(pickled_X_pred, X_pred)
 
 
-def check_estimators_partial_fit_n_features(name, Alg):
-    # check if number of features changes between calls to partial_fit.
-    if not hasattr(Alg, 'partial_fit'):
+def check_partial_fit_n_features(name, Estimator):
+    """Check if ValueError is raised if number of features changes between
+    multiple partial_fit calls.
+    """
+
+    if not hasattr(Estimator, 'partial_fit'):
         return
+
     X, y = make_blobs(n_samples=50, random_state=1)
     X -= X.min()
+
     with warnings.catch_warnings(record=True):
-        alg = Alg()
-    set_fast_parameters(alg)
-    if isinstance(alg, ClassifierMixin):
+        estimator = Estimator()
+
+    set_random_state(estimator)
+    set_fast_parameters(estimator)
+
+    if isinstance(estimator, ClassifierMixin):
         classes = np.unique(y)
-        alg.partial_fit(X, y, classes=classes)
+        estimator.partial_fit(X, y, classes=classes)
+        assert_raises(ValueError, estimator.partial_fit, X[:, :-1], y)
+
+    elif isinstance(estimator, RegressorMixin):
+        estimator.partial_fit(X, y)
+        assert_raises(ValueError, estimator.partial_fit, X[:, :-1], y)
+
+    elif isinstance(estimator, TransformerMixin):
+        estimator.partial_fit(X)
+        assert_raises(ValueError, estimator.partial_fit, X[:, :-1])
+
+
+def check_partial_fit_general(name, Estimator):
+    """General tests for partial_fit method"""
+
+    X, y = make_blobs(n_samples=50, random_state=0)
+    X -= X.min()
+
+    test_X, test_y = make_blobs(n_samples=50, random_state=1)
+    test_X -= test_X.min()
+
+    with warnings.catch_warnings(record=True):
+        estimator = Estimator()
+
+    set_random_state(estimator)
+    set_fast_parameters(estimator)
+
+    cloned = clone(estimator)
+
+    msg = ("partial_fit of the clone of estimator %s returned a different"
+           " model" % name)
+
+    if isinstance(estimator, ClassifierMixin):
```

Comment:
Sure thanks for the suggestion... :) 
I'll add a separate commit and lets see if any one has a different opinion on that...